src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/CollectionServices.cs

```diff
@@ -144,11 +145,13 @@ private static List<T> FastAppendToListAllowNulls<T>(this List<T>? source, T val
         }
 
         public static List<T>? FastAppendToListAllowNulls<T>(
-                        this List<T>? source, T value,
+                        this List<T>? source, T? value,
                         IEnumerable<T>? second)
+            where T : class
         {
             if (second == null)
             {
+                Debug.Assert(value is object);
```

Comment:
How do we know this assert is valid?

Reply:
The two callers get their values from `ComposablePartDefinition.TryGetExports`. That method can return with 
- both `singleMatch` and `multipleMatches` being null (`return false;`, but then `FastAppendToListAllowNulls` won't be called because of calling pattern, copied below), 
- or `multipleMatches` being set and returning true, 
- or `singleMatch` being set and returning true.

`singleMatch` corresponds to parameter `value` and `multipleMatches` corresponds to parameter `second` here.

Calling pattern:
``` csharp
                    if (part.TryGetExports(definition, out Tuple<ComposablePartDefinition, ExportDefinition>? singleMatch, out IEnumerable<Tuple<ComposablePartDefinition, ExportDefinition>>? multipleMatches))
                    {
                        exports = exports.FastAppendToListAllowNulls(singleMatch, multipleMatches);
                    }
```

---
In reply to: [378614963](https://github.com/dotnet/runtime/pull/32090#discussion_r378614963) [](ancestors = 378614963)